### PR TITLE
[Distributed][Example] Distributed training for unsupervised graphsage

### PR DESCRIPTION
**Pull request by classicsong**
## Description
Distributed training example for unsupervised graphsage

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [ ] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: examples/pytorch/graphsage/experimental/README.md [hunk 1]
```diff
@@ -19,6 +19,11 @@ the number of nodes, the number of edges and the number of labelled nodes.
 python3 partition_graph.py --dataset ogb-product --num_parts 4 --balance_train --balance_edges
 ```
 
+```bash
+# partitoin for unsupervised learning
+python3 partition_graph.py --dataset ogb-product --num_parts 4 --balance_train --balance_edges --num_hops 2
```
Comment (zheng-da) [inline examples/pytorch/graphsage/experimental/README.md:24]: why 2 hops?

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+
+        # Maintain the correspondence between heads, tails and negative tails as two
+        # graphs.
+        # pos_graph contains the correspondence between each head and its positive tail.
+        # neg_graph contains the correspondence between each head and its negative tails.
+        # Both pos_graph and neg_graph are first constructed with the same node space as
+        # the original graph.  Then they are compacted together with dgl.compact_graphs.
+        pos_graph = dgl.graph((heads, tails), num_nodes=self.g.number_of_nodes())
+        neg_graph = dgl.graph((neg_heads, neg_tails), num_nodes=self.g.number_of_nodes())
+        pos_graph, neg_graph = dgl.compact_graphs([pos_graph, neg_graph])
```
Comment (zheng-da) [inline examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:59]: this is to compact two graphs together so that the same nodes will have the same node Ids?

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+
+        seeds = pos_graph.ndata[dgl.NID]
```
Comment (zheng-da) [inline examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:61]: you only get node Ids from the pos graph. both graphs have the same nodes?

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+        blocks = []
+        for fanout in self.fanouts:
+            # For each seed node, sample ``fanout`` neighbors.
+            frontier = self.sample_neighbors(self.g, seeds, fanout, replace=True)
+            # Remove all edges between heads and tails, as well as heads and neg_tails.
+            _, _, edge_ids = frontier.edge_ids(
+                th.cat([heads, tails, neg_heads, neg_tails]),
+                th.cat([tails, heads, neg_tails, neg_heads]),
+                return_uv=True)
```
Comment (zheng-da) [inline examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:70]: concatenate them in advance?

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py [hunk 5]
```diff
@@ -0,0 +1,393 @@
...
+        return y
+
+def load_subtensor(g, input_nodes, device):
+    """
+    Copys features and labels of a set of nodes onto GPU.
+    """
+    batch_inputs = g.ndata['features'][input_nodes].to(device)
+    return batch_inputs
+
+class CrossEntropyLoss(nn.Module):
```
Comment (zheng-da) [inline examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:167]: why is this a class? it's kind of weird to define it as a nn module.

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+    def forward(self, block_outputs, pos_graph, neg_graph):
+        with pos_graph.local_scope():
+            pos_graph.ndata['h'] = block_outputs
+            pos_graph.apply_edges(fn.u_dot_v('h', 'h', 'score'))
+            pos_score = pos_graph.edata['score']
+        with neg_graph.local_scope():
+            neg_graph.ndata['h'] = block_outputs
+            neg_graph.apply_edges(fn.u_dot_v('h', 'h', 'score'))
+            neg_score = neg_graph.edata['score']
```
Comment (zheng-da) [inline examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:176]: why do you right these two in a graph way? tensor operations will be more intuitive.

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+    model.eval()
+    with th.no_grad():
+        pred = model.inference(g, inputs, batch_size, device)
+
+    return pred
+
+def compute_acc(emb, labels, train_nids, val_nids, test_nids):
+    """
+    Compute the accuracy of prediction given the labels.
+    """
```
Comment (zheng-da) [inline examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:200]: can you comment on what task it is?

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+    f1_micro_test = skm.f1_score(labels[test_nids], pred[test_nids], average='micro')
+    f1_macro_eval = skm.f1_score(labels[val_nids], pred[val_nids], average='macro')
+    f1_macro_test = skm.f1_score(labels[test_nids], pred[test_nids], average='macro')
+    return f1_micro_eval, f1_micro_test, f1_macro_eval, f1_macro_test
+
+def run(args, device, data):
+    # Unpack data
+    train_eids, train_nids, in_feats, g, global_train_nid, global_valid_nid, global_test_nid, labels = data
+    # Create sampler
+    sampler = NeighborSampler(g, [int(fanout) for fanout in args.fan_out.split(',')], train_nids,
```
Comment (zheng-da) [inline examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:223]: this is weird. the negative edges are corrected by the nodes in `train_nids`? why are they corrupted by the nodes in the training set instead of all nodes in the local partition?

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py [hunk 9]
```diff
@@ -0,0 +1,393 @@
...
+
+def main(args):
+    if not args.standalone:
+        th.distributed.init_process_group(backend='gloo')
+    g = dgl.distributed.DistGraph(args.ip_config, args.graph_name, conf_file=args.conf_path)
+    print('rank:', g.rank())
+    print('number of edges', g.number_of_edges())
+
+    train_eids = dgl.distributed.edge_split(th.arange(g.number_of_edges()), g.get_partition_book(), force_even=True)
+    train_nids = dgl.distributed.node_split(th.arange(g.number_of_nodes()), g.get_partition_book())
```
Comment (zheng-da) [inline examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:349]: this is incorrect. the input of `node_split` and `edge_split` should be a boolean mask array, instead of a Id vector. I'll update the API to support both.

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+    global_train_nid = g.ndata['train_mask'][np.arange(g.number_of_nodes())]
+    global_valid_nid = g.ndata['val_mask'][np.arange(g.number_of_nodes())]
+    global_test_nid = g.ndata['test_mask'][np.arange(g.number_of_nodes())]
```
Comment (zheng-da) [inline examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:352]: you get all nodes in the training set?

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py [hunk 11]
```diff
@@ -0,0 +1,393 @@
...
+        epoch += 1
+
+        if args.standalone:
+            pred = generate_emb(model,  g, g.ndata['features'], args.batch_size_eval, device)
+        else:
+            pred = generate_emb(model.module, g, g.ndata['features'], args.batch_size_eval, device)
+        if g.rank() == 0:
+            f1_micro_eval, f1_micro_test, f1_macro_eval, f1_macro_test = compute_acc(pred, labels, global_train_nid, global_valid_nid, global_test_nid)
+            print('f1 micro eval {:.4f} f1 micro test {:.4f}, f1 macro eval {:.4f} f1 macro test {:.4f}'.format(
+                f1_micro_eval, f1_micro_test, f1_macro_eval, f1_macro_test))
```
Comment (zheng-da) [inline examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:322]: this is to use node classification to validate the embeddings? this is weird to me. in the true unsupervised setting, we cannot validate the embeddings with a node classification task. otherwise, why do we not train the embeddings with the node classification task? second, it doesn't seem your implementation is very scalable.

File: python/dgl/distributed/dist_graph.py [hunk 12]
```diff
@@ -453,6 +454,25 @@ def rank(self):
...
+            The edge ID array.
+
+        Returns
+        -------
+        tensor
+            The source node ID array.
+        tensor
+            The destination node ID array.
+        """
+        return find_edges(self, edges)
```
Comment (zheng-da) [inline python/dgl/distributed/dist_graph.py:474]: you can rename the one imported graph_services with another name? it looks like recursive call.

Comment (zheng-da) [reply python/dgl/distributed/dist_graph.py:474]: also please provide a unit test for this operation.

File: examples/pytorch/graphsage/experimental/README.md (hunk 1 above)
Comment (classicsong) [reply examples/pytorch/graphsage/experimental/README.md:24]: I just copy the config from train_sampling_unsupervised.
It is just a hyperparameter

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py (hunk 2 above)
Comment (classicsong) [reply examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:59]: Yes

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py (hunk 3 above)
Comment (classicsong) [reply examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:61]: Yes, they are compacted.

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py (hunk 4 above)
Comment (classicsong) [reply examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:70]: This is for remove edges.
We can have a flag to control whether to do remove edges.

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py (hunk 10 above)
Comment (classicsong) [reply examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:352]: For validation

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py (hunk 11 above)
Comment (classicsong) [reply examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:322]: For validation it uses node classification to validate the embeddings.
It feed the embeddings to LogisticRegression

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py (hunk 9 above)
Comment (classicsong) [reply examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:349]: I c. I will use th.ones to make the mask.

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py (hunk 7 above)
Comment (classicsong) [reply examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:200]: Sure

File: examples/pytorch/graphsage/experimental/README.md (hunk 1 above)
Comment (zheng-da) [reply examples/pytorch/graphsage/experimental/README.md:24]: but this is partition_graph.py. i don't think train_sampling_unsupervised invokes this script.

File: tests/distributed/test_distributed_sampling.py [hunk 13]
```diff
@@ -220,3 +256,5 @@ def test_rpc_in_subgraph():
         check_rpc_sampling_shuffle(Path(tmpdirname), 2)
         check_rpc_sampling(Path(tmpdirname), 2)
         check_rpc_sampling(Path(tmpdirname), 1)
+        check_rpc_find_edges(Path(tmpdirname), 2)
+        check_rpc_find_edges(Path(tmpdirname), 1)
```
Comment (zheng-da) [inline tests/distributed/test_distributed_sampling.py:260]: is the test here the same as the one in test_dist_graph_store?

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py [hunk 14]
```diff
@@ -0,0 +1,411 @@
...
+
+    if not args.standalone:
+        g._client.barrier()
+
+        # save features into file
+        if g.rank() == 0:
+            th.save(pred, 'emb.pt')
+
+        dgl.distributed.shutdown_servers()
+        dgl.distributed.finalize_client()
```
Comment (zheng-da) [inline examples/pytorch/graphsage/experimental/train_dist_unsupervised.py:346]: i don't think we need to close RPC explicitly now.

File: tests/distributed/test_distributed_sampling.py (hunk 13 above)
Comment (classicsong) [reply tests/distributed/test_distributed_sampling.py:260]: 2 will create partitions as local and remote.

File: examples/pytorch/graphsage/experimental/README.md (hunk 1 above)
Comment (classicsong) [reply examples/pytorch/graphsage/experimental/README.md]: removed

File: examples/pytorch/graphsage/experimental/train_dist_unsupervised.py (hunk 14 above)
Comment (classicsong) [reply examples/pytorch/graphsage/experimental/train_dist_unsupervised.py]: removed
